Server/src/BaseQuasarServer.cpp

```diff
@@ -198,6 +204,73 @@ std::string BaseQuasarServer::getApplicationPath() const
     return std::string(serverSettingsPath);
 }
 
+std::string BaseQuasarServer::getProcessEnvironmentVariables() const
+{
+    std::ostringstream result;
+#ifdef __linux__
+    extern char **environ;
+    char *s = *environ;
+    for (int i = 0; s; ++i)
+    {
+        result << s << std::endl;
+        s = *(environ + i);
+    }
+#elif _WIN32
+    auto envBlockDeleterFn = [](LPTCH p) { FreeEnvironmentStrings(p); };
+    auto envBlock = std::unique_ptr<TCHAR, decltype(envBlockDeleterFn)>{ GetEnvironmentStrings(), envBlockDeleterFn };
+    // envBlock format: "keyA=valA\0keyB=valB\0...keyXXX=valXXX\0\0". Double \0\0 denotes end of block
+    for (TCHAR* pos = envBlock.get(); pos != nullptr && *pos != TCHAR('\0'); )
+    {
+        const std::string envVar(pos); // capture from pos up to '\0'
+        result << envVar << std::endl;
+        pos += envVar.length() + 1; // +1 ? wind pos past '\0' at end of current key/val pair
+    }
+#endif
+    return result.str();
+}
+
+/**
+* Returns string in format: logged in user (effective process owner).
+* Examples, where a user 'quasar' is logged in...
+* == posix ==
+* 'quasar' runs process regular - returns 'quasar(quasar)'
+* 'quasar' runs process with sudo - returns 'quasar(root)'
+* == windows ==
+* 'quasar' runs process regular - returns 'quasar(normal)'
+* 'quasar' runs process with admin rights (or admin shell) 'quasar(elevated)'
+*/
+std::string BaseQuasarServer::getProcessOwner() const
+{
+    std::ostringstream result;
+#ifdef __linux__
+    const auto userID = getlogin();
+    result << (userID != nullptr ? userID : "unknown");
+
+    std::string effectiveUserID = "unknown";
```

Comment:
Please clarify: in Linux we have a distinction between UID and effective UID; the functions to get these are getuid() and geteuid(). I have rough time figuring out why it's considered that getuid() (in the submitted code) is considered "effective" while the actual effective (geteuid()) is not taken into account at all?

Reply:
Yep, well spotted: will change getuid to geteuid for effective account.

== for info ==
Sadly, for logged in user, getuid is not the guy (returns root if run as sudo).
`void tryThings()
{ 
  std::ostringstream result;
  
  auto getUserName = [](const uid_t& uid){
    const auto pswd = getpwuid(uid);
    const std::string name(pswd != nullptr && pswd->pw_name != nullptr ? pswd->pw_name : "unknown");
    std::ostringstream s;
    s << uid << ":" << name;
    return s.str();
  };
  
  result << getUserName(getuid()) << "(" << getUserName(geteuid()) << ")";
  std::cout << __FUNCTION__ << " " <<result.str() << std::endl;
}`

1. _bash-4.2$ ./process_owner_
`tryThings 56850:bfarnham(56850:bfarnham)`
Great!

2. _bash-4.2$ sudo ./process_owner_
`tryThings 0:root(0:root)`
From debugging/diagnosing POV I'd like to know the logged in user as well as effective process owner.
Which seems less useful than the info obtained from the current implementation, where we report on the logged in user (bfarnham) and the

Reply:
Hi @ben-farnham ,
I still think that using getlogin + geteuid is not enough. 
We should have output of either getuid + geteuid, or if you prefer getlogin + getuid + geteuid.

Reasons are that getlogin is weak. On the other hand info getuid and geteuid comes directly from the kernel.

Read this: 

>      Unfortunately, it is often rather easy to fool getlogin().  Sometimes it does not work at all, because some program > messed up the utmp file.  Often, it gives only the first 8 characters of the login
>       name.  The user currently logged in on the controlling terminal of our program need not be the user who started > it.  Avoid getlogin() for security-related purposes.

So in a world where you have: uid, euid, SUID executable bit, caps, distirbuted permissions and ACLs I think uid + euid are mandatory, getlogin() is OK to have but to me optional and might be actually misleading (see the paragraph cited above).
Note that sudo has some specific mode of operation which is different from having SUID bit set (u+sx mode); both will give you elevated rights but completely differently.

Reply:
fair enough: will publish
getlogin + uid + eiud

Reply:
Now looks like (linux, run as sudo)
Process owner: login:bfarnham uid:0,root euid: 0,root